paperqa/agents/task.py

```diff
@@ -487,3 +493,229 @@ def __len__(self) -> int:
+            "human_answer": human_answer,
+            "pqa_answer_index": pqa_answer_index,
+            "winner": winner,
+            "llm_response": result.text,
+            "paper_search_ids": paper_search_ids,
+            "pqa_context": pqa_context,
+            "gt_doc_ids": gt_doc_ids,
+            "were_gt_docs_found": were_gt_docs_found,
+        }
+
+        results_dir = os.path.join(
+            "data/rag-qa-benchmarking", f"results_{llm_model_name}"
+        )
+        os.makedirs(results_dir, exist_ok=True)
+        json_path = os.path.join(results_dir, f"{qid}.json")
+        with open(json_path, "w") as f:
+            json.dump(evaluation_results, f, indent=2)
+
+    async def pairwise_evaluation(
+        self, qid: str | UUID, question: str, pqa_answer: str, human_answer: str
+    ) -> float:
+
+        paper_search_ids = [int(doc.docname) for doc in self.state.docs.docs.values()]
+
+        pairwise_eval_llm = LiteLLMModel(name=self.pairwise_eval_llm)
+        pqa_answer = strip_citations(pqa_answer)
+
+        pqa_answer_index = 1 if random.random() < 0.5 else 2  # noqa: PLR2004
+        data = {
+            "question": question,
+            "answer1": pqa_answer if pqa_answer_index == 1 else human_answer,
+            "answer2": human_answer if pqa_answer_index == 1 else pqa_answer,
+        }
+
+        result = await pairwise_eval_llm.run_prompt(
+            prompt=lfrqa_prompt_template,
+            data=data,
+            system_prompt=lfrqa_system_prompt,
+        )
+
+        best_answer_index = self.extract_best_answer_index(result.text)
+        winner = (
+            "paperqa"
+            if best_answer_index == pqa_answer_index
+            else "human" if best_answer_index != 0 else "tie"
+        )
+
+        self.log_results_to_json(
+            self._settings.llm,
+            qid,
+            question,
+            pqa_answer,
+            paper_search_ids,
+            self.state.session.context,
+            human_answer,
+            self.gt_doc_ids,
+            pqa_answer_index,
+            winner,
+            result,
+        )
+
+        return (
+            self._rewards["win"]
+            if winner == "paperqa"
+            else self._rewards["lose"] if winner == "human" else self._rewards["tie"]
+        )
+
+    async def step(
+        self, action: ToolRequestMessage
+    ) -> tuple[Messages, float, bool, bool]:
+        messages, reward, done, truncated = await super().step(action)
+        if done:
+            reward = await self.pairwise_evaluation(
+                self.qid, self.question, self.state.session.answer, self.human_answer
+            )
+        return messages, reward, done, truncated
+
+
+class LFRQATaskDataset(
+    TaskDataset[GradablePaperQAEnvironment], ComputeTrajectoryMetricsMixin
+):
+    """Task dataset for custom evaluation of non-multiple choice questions."""
+
+    def __init__(
+        self,
+        data_path: str,
```

Comment:
Can you have this data_path be optional, and have the code autodownload the data and cache in `pqa_directory("datasets")`